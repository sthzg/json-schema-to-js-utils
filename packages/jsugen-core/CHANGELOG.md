# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.1.3](https://github.com/sthzg/jsugen/compare/v0.1.2...v0.1.3) (2019-05-01)

**Note:** Version bump only for package @sthzg/jsugen-core





## [0.1.2](https://github.com/sthzg/jsugen/compare/v0.1.1...v0.1.2) (2019-04-27)


### Bug Fixes

* Fix prettier script in root package.json ([0e1af61](https://github.com/sthzg/jsugen/commit/0e1af61))





## [0.1.1](https://github.com/sthzg/jsugen/compare/v0.1.0...v0.1.1) (2019-02-28)


### Bug Fixes

* update parser setting to fix deprecation message ([11a3f5e](https://github.com/sthzg/jsugen/commit/11a3f5e))





# [0.1.0](https://github.com/sthzg/jsugen/compare/v0.0.4...v0.1.0) (2019-02-28)

**Note:** Version bump only for package @sthzg/jsugen-core





## [0.0.4](https://github.com/sthzg/jsugen/compare/v0.0.3...v0.0.4) (2019-02-28)

**Note:** Version bump only for package @sthzg/jsugen-core





## [0.0.3](https://github.com/sthzg/jsugen/compare/v0.0.3-beta.1...v0.0.3) (2019-02-28)


### Bug Fixes

* make build chain work and publish ([fc00c4e](https://github.com/sthzg/jsugen/commit/fc00c4e))
