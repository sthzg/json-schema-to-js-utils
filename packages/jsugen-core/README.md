# `jsugen-core`

> This package contains common logic reused in various sub projects.
