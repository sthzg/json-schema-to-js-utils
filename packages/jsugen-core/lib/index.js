export {
  COMMA,
  DEFAULT_GET_IMPORT,
  DEFAULT_FILE_DOCSTRING,
  DEFAULT_PRETTIER_OPTIONS,
  DRY_RUN_BANNER,
  EMPTY_STRING,
  ENCODING,
  POSTFIX,
  PREFIX,
} from './constants';
export { ModuleFormat } from './enums';
export { withCompileToTemplate } from './withCompileToTemplate';
export { toTemplateRawStringReducer, withPrettier, withWrite } from './output';
export {
  enrichWithPathNodeVars,
  addMemberDefinitionsForNonEnumArrayIndexes,
} from './enrichment/pathNodeTemplateVars';
export { generate } from './main';
export { WriteConfig } from './models';
export { parseSource } from './parsing';
export {
  byIsLeaf,
  byMemberDefinitionIsEnum,
  byMemberDefinitionMemberName,
  byMemberDefinitionPathNodes,
  byPathNodeIsArray,
  byPathNodeIsEnum,
  byPathNodesEnumValues,
  byPathNodesIsArrayLeaf,
  byPathNodesIsEnumLeaf,
  byPathNodeType,
  byTemplateRaw,
} from './selectors';
export {
  /* array.js */
  arrayToEnum,
  maybeItem,
  /* object.js */
  enrichIn,
  enrichInData,
  enrichInTemplate,
  mergeConcat,
  mergeSet,
  /* format.js */
  toCamelCase,
  toUpperFirstCamelCase,
  toUpperSnakeCase,
  wrapInCurlyBraces,
  wrapInSingleQuote,
  wrapInRoundBrackets,
  /* number.js */
  minusOne,
  plusOne,
  /* rxjs.js */
  reduceGroupToArray,
  /* log.js */
  log,
  logError,
  /* string.js */
  withPrependToString,
} from './utils';
