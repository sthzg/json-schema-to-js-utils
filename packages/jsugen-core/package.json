{
  "name": "@sthzg/jsugen-core",
  "version": "0.13.0",
  "description": "> TODO: description",
  "author": "Stephan Herzog <sthzgvie@gmail.com>",
  "homepage": "https://github.com/sthzg/jsugen#readme",
  "license": "MIT",
  "main": "lib/index.js",
  "publishConfig": {
    "access": "public"
  },
  "directories": {
    "lib": "lib"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/sthzg/jsugen.git"
  },
  "scripts": {
    "test": "echo \"Error: run tests from root\" && exit 1"
  },
  "bugs": {
    "url": "https://github.com/sthzg/jsugen/issues"
  },
  "dependencies": {
    "@cloudflare/json-schema-walker": "^0.1.1",
    "@sthzg/jsugen-sources": "^0.13.0",
    "es6-error": "^4.1.1",
    "fs-extra": "^8.0.1",
    "glob": "^7.1.4",
    "json-schema-ref-parser": "^7.1.0",
    "prettier": "^1.17.0",
    "rxjs": "^6.5.1"
  },
  "peerDependencies": {
    "lodash-es": ">=4"
  }
}
