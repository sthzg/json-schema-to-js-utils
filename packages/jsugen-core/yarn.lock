# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


"@cloudflare/json-schema-walker@^0.1.1":
  version "0.1.1"
  resolved "https://registry.yarnpkg.com/@cloudflare/json-schema-walker/-/json-schema-walker-0.1.1.tgz#d1cc94065327b0b3800158db40cb78124a3476de"
  integrity sha512-P3n0hEgk1m6uKWgL4Yb1owzXVG4pM70G4kRnDQxZXiVvfCRtaqiHu+ZRiRPzmwGBiLTO1LWc2yR1M8oz0YkXww==

lodash.castarray@^4.4.0:
  version "4.4.0"
  resolved "https://registry.yarnpkg.com/lodash.castarray/-/lodash.castarray-4.4.0.tgz#c02513515e309daddd4c24c60cfddcf5976d9115"
  integrity sha1-wCUTUV4wna3dTCTGDP3c9ZdtkRU=

lodash.first@^3.0.0:
  version "3.0.0"
  resolved "https://registry.yarnpkg.com/lodash.first/-/lodash.first-3.0.0.tgz#5dae180d7f818ee65fc5b210b104a7bbef98a16a"
  integrity sha1-Xa4YDX+BjuZfxbIQsQSnu++YoWo=

lodash.isempty@^4.4.0:
  version "4.4.0"
  resolved "https://registry.yarnpkg.com/lodash.isempty/-/lodash.isempty-4.4.0.tgz#6f86cbedd8be4ec987be9aaf33c9684db1b31e7e"
  integrity sha1-b4bL7di+TsmHvpqvM8loTbGzHn4=

lodash.merge@^4.6.1:
  version "4.6.1"
  resolved "https://registry.yarnpkg.com/lodash.merge/-/lodash.merge-4.6.1.tgz#adc25d9cb99b9391c59624f379fbba60d7111d54"
  integrity sha512-AOYza4+Hf5z1/0Hztxpm2/xiPZgi/cjMqdnKTUWTBSKchJlxXXuUSxCCl8rJlf4g6yww/j6mA8nC8Hw/EZWxKQ==

lodash.snakecase@^4.1.1:
  version "4.1.1"
  resolved "https://registry.yarnpkg.com/lodash.snakecase/-/lodash.snakecase-4.1.1.tgz#39d714a35357147837aefd64b5dcbb16becd8f8d"
  integrity sha1-OdcUo1NXFHg3rv1ktdy7Fr7Nj40=

prettier@^1.15.3:
  version "1.15.3"
  resolved "https://registry.yarnpkg.com/prettier/-/prettier-1.15.3.tgz#1feaac5bdd181237b54dbe65d874e02a1472786a"
  integrity sha512-gAU9AGAPMaKb3NNSUUuhhFAS7SCO4ALTN4nRIn6PJ075Qd28Yn2Ig2ahEJWdJwJmlEBTUfC7mMUSFy8MwsOCfg==
