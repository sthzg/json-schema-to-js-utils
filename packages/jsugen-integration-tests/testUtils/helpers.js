import path from 'path';
import pkgDir from 'pkg-dir';
import { WriteConfig } from '@sthzg/jsugen-core';
import { NEWLINE } from '@sthzg/jsugen-core/lib/constants';

// ---
// Constants.
// ---

/**
 * Target directory name for file writes during integration tests.
 */
export const OUT_DIR_NAME = 'out';

/**
 * Dry run configuration settings for integration tests.
 */
export const DryRunWriteConfig = new WriteConfig({
  dryRun: true,
  silent: true,
});

// ---
// Paths.
// ---

export const TEST_RESOURCES_DIR = path.resolve(__dirname, '../resources');
export const EXAMPLE_SCHEMA_PATH = path.join(
  TEST_RESOURCES_DIR,
  'example.schema.json',
);
export const CONTACT_SCHEMA_PATH = path.join(
  TEST_RESOURCES_DIR,
  'example.contact.json',
);
export const ARRAY_OF_OBJECTS_SCHEMA_PATH = path.join(
  TEST_RESOURCES_DIR,
  'example.array-of-objects.schema.json',
);
export const EMPTY_SCHEMA_PATH = path.join(
  TEST_RESOURCES_DIR,
  'empty.schema.json',
);

// ---
// Utils.
// ---

export function byOutDirectory(target) {
  return path.resolve(pkgDir.sync(__filename), OUT_DIR_NAME, target);
}

export function withSnapshotCheck(done, annotation) {
  const chunks = [];

  return {
    next(output) {
      chunks.push(output);
    },
    complete() {
      const content = chunks.join(NEWLINE);
      expect(content).toMatchSnapshot(annotation);
      done();
    },
  };
}
