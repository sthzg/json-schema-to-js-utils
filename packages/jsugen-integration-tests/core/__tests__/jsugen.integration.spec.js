import fs from 'fs-extra';
import { generate } from '@sthzg/jsugen-core';
import path from 'path';
import { byOutDirectory } from '../../testUtils/helpers';

/**
 * Set to `true` to to keep the generated files in `OUT_DIR`.
 */
const SKIP_DELETE = false;

const OUT_DIR = byOutDirectory('core');
const byPathInOut = needle => path.resolve(__dirname, OUT_DIR, needle);
const existsInOut = needle => fs.existsSync(byPathInOut(needle));

afterEach(() => {
  if (!SKIP_DELETE) {
    fs.emptyDirSync(OUT_DIR);
  }
});

// ---
// Dry Run Tests.
// ---
describe('main: dry run', () => {
  const config = {
    dryRun: true,
    definitions: [
      {
        files: ['**/globTesting/**/*.schema.json'],
        generators: [
          '@sthzg/jsugen-generate-selector-functions',
          '@sthzg/jsugen-generate-lodash-object-paths',
        ],
        output: {
          baseDirectory: OUT_DIR,
        },
      },
    ],
  };

  test('should glob the expected files and run the configured generators', done => {
    expect.assertions(4);

    generate({ config }).subscribe(output => {
      expect(output).toMatchSnapshot();
      done();
    });
  });
});

// ---
// IO Tests: Static output directory configuration.
// ---
describe('main: static output directory', () => {
  const config = {
    dryRun: false,
    definitions: [
      {
        files: ['**/globTesting/**/*.schema.json'],
        generators: ['@sthzg/jsugen-generate-selector-functions'],
        output: {
          baseDirectory: OUT_DIR,
        },
      },
    ],
  };

  test('should generate the files at the expected location for output mode: static', done => {
    expect.assertions(2);

    generate({ config }).subscribe({
      complete() {
        expect(existsInOut('Hello/selectorFunctions.js')).toBe(true);
        expect(existsInOut('World/selectorFunctions.js')).toBe(true);

        done();
      },
    });
  });
});

// ---
// IO Tests: Source-file relative output directory configuration.
// ---
describe('main: colocated writes', () => {
  const config = {
    dryRun: false,
    definitions: [
      {
        files: ['**/colocatedOutput/**/*.schema.json'],
        generators: ['@sthzg/jsugen-generate-selector-functions'],
        output: {
          baseDirectory:
            '<sourceFile>../../../out/core/relative-to-source-file',
        },
      },
    ],
  };

  test('should generate the files at the expected location for output mode: relative', done => {
    expect.assertions(2);

    generate({ config }).subscribe({
      complete() {
        expect(
          existsInOut('relative-to-source-file/Hello/selectorFunctions.js'),
        ).toBe(true);
        expect(
          existsInOut('relative-to-source-file/World/selectorFunctions.js'),
        ).toBe(true);

        done();
      },
    });
  });
});
