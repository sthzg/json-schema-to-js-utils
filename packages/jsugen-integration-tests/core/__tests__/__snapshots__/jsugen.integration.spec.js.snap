// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main: dry run should glob the expected files and run the configured generators 1`] = `
"/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

import get from 'lodash-es/get';

export function byHello(property, defaultReturn) {
  return get(property, ['hello'], defaultReturn);
}

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

export const byHelloPath = () => ['hello'];

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

import get from 'lodash-es/get';

export function byWorld(property, defaultReturn) {
  return get(property, ['world'], defaultReturn);
}

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

export const byWorldPath = () => ['world'];
"
`;

exports[`main: dry run should respect the config.imports.lodashGet setting: expect the customized import statement 1`] = `
"/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

import { get } from 'lodash';

export function byHello(property, defaultReturn) {
  return get(property, ['hello'], defaultReturn);
}

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

export const byHelloPath = () => ['hello'];

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

import { get } from 'lodash';

export function byWorld(property, defaultReturn) {
  return get(property, ['world'], defaultReturn);
}

/**
 * @file This file is auto-generated. Do not modify its contents manually.
 */

export const byWorldPath = () => ['world'];
"
`;
